### DISCUSSION TITLE
Would Klicker be interested in being optionally stateless? / reducing deployment footprint?

### DISCUSSION THREAD

**Discussion by jamesread** [uzh-bf/klicker-react#2544] [Q&A] [answered]
Hi there,

Klicker looks like a great project. I've been searching for something like this for a long time. Big advantages for me is that it's AGPL, JavaScript, and I can see it's actively developed too.

However, I ran into some questions around the design goals of the project. I am interested if the developers would consider making it more stateless, autoconfigurable and allow configuration via YAML (or some sort of text file configuration). 

I tried deploying Klicker, but found it frustrating that it needs MonogoDB, optionally redis, and a few environment variables set like the JWT token. The effort to "just try it out" means I had to set up a few things first. I had to search through the docs to find the environment variables to set. I could not get the MongoDB connection working, and I gave up in the end.

1. Stateless - Would the project be interested in making MongoDB optional in the future (eg, just save "results" in memory)? 
2. Autoconfiguration defaults - Would the project also think about some sort of "autoconfiguration defaults"? For example, using a UUID for the JWT token if it's not set? I host many presentations, and I'd really like something that I can deploy/destroy with basically zero configuration. 
3. Config in a text file - so I could easily load them into the tool. This makes it easy to copy/paste questions between sessions.

If the project is open to these ideas then I'd be really interested in collaborating and contributing some changes. If not, I think I'll end up writing my own tool as these design goals are important to me - but I'm here to ask first! :-) 

Thanks for taking the time to read, and a great looking project.

**Comment by rschlaefli** [comment, marked as answer]
Hi @jamesread,

Thanks a lot for your interest in our project!

I am sorry that you were frustrated by our technical setup. Our documentation has not been great in terms of information on self-hosting, and most of our limited resources went into the development of new features.

That being said, I would be very open to improving the setup experience for a self-hosted KlickerUZH. Our goal has always been to make the project open and accessible for external users as well. We have already taken some important steps with our latest release, with a new Helm chart for Kubernetes deployments, and improved default configuration with Docker Compose for a local deployment. Our documentation does not yet reflect these changes, but we will be revamping the documentation for this in October.

As for the specific points you mentioned: the main design goal in terms of deployment has been to make the KlickerUZH easy to host anywhere using Docker containers. Our goal would be that it would be very easy for you to host a local instance with Docker Compose, or to host on any cloud with Kubernetes or another orchestrator of your choice (hence our new Helm chart).

1. Stateless: would there be any point to be made against using Docker Compose when running your own instance, and an ephemeral or local volume for the MongoDB and Redis instance? From a technical point of view, we have deliberately chosen the MongoDB/Redis stack (e.g., using Redis for incoming responses from participants, to quickly process the responses in our large classes with 700+ students), and it would be non-trivial to make this "pluggable". We could certainly investigate this further, and you probably have some good ideas for that (I have not given it that much thought up to now). What we should probably make pluggable eventually is not necessarily having to use S3 to store images (e.g., not having to use Minio in a local deployment or a cloud like AWS). However, all of this is not currently on our roadmap, so would need further discussion.

2. Autoconfiguration defaults: there are some important improvements that we will still need to make, before the KlickerUZH is "autoconfigurable" and can just be started and used immediately (but this is definitely the goal). ATM you need to create your first user yourself, need to create a bucket for image assets that belong to questions in the local Minio instance, and need to set several environment variables like the APP_SECRET. Performing this initialization with some bootstrapping procedure (e.g., autogeneration of a first admin user, the app secret, etc.) would be a very good idea, and we are certainly open to adding this.

3. Not sure I understood correctly. But I gather you would like to have the questions stored in a text file, and have the KlickerUZH load them on startup, so you can then create sessions? Might be possible to do when using a bootstrapping script, but maybe it would be easier if the Docker volume of MongoDB/Redis is persisted, and the KlickerUZH can be easily started again the next time it is used? Feel free to correct me if this is not what you meant.

Anyways, I think we would need some more information on your specific use case and how you would like to use the KlickerUZH, but we are certainly open to improvements in the area of running an actual instance.

Thanks again for your interest in the KlickerUZH!

Best,
Roland

**Comment by rschlaefli** [comment]
Hi @jamesread, just wondering whether you have found any solution to the specific use case you described here. We might pick up some of these issues again in the future, and it would be interesting to know what the outcome was in your case. We have simplified local setup quite a bit (docker-compose now normally “just works”) but we are still a long way from plug-and-play for a small local production server.

**Reply by jamesread** [reply]
Hey @rschlaefli , thanks for reaching out, I am very much still interested, just very limited on free time. I wanted to comment here that I've not just ignored this message, but free time to add anything meaningful is tricky. I'll check back here and add something meaningful when I can :-)
